File: 2018 ICPC Vietnam National Programming Contest/gottacatchemall.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Program {
    static void Main() {
        int n = int.Parse(Console.ReadLine());
        
        List<(int, int)> junctions = new List<(int, int)>();
        for (int i = 0; i < n - 1; i++) {
            int[] edge = Console.ReadLine().Split().Select(int.Parse).ToArray();
            int u = edge[0], v = edge[1];
            if (u > v) (u, v) = (v, u);
            junctions.Add((u, v));
        }
        
        junctions = junctions.Distinct().OrderBy(x => x.Item1).ThenBy(x => x.Item2).ToList();
        
        int[] count = new int[n + 1];
        foreach (var (u, v) in junctions) {
            count[u]++;
            count[v]++;
        }
        
        List<int>[] connected = new List<int>[n + 1];
        for (int i = 1; i <= n; i++) {
            connected[i] = new List<int>();
        }
        
        foreach (var (u, v) in junctions) {
            if (count[u] != count[v] ? count[u] > count[v] : u < v) {
                connected[u].Add(v);
            } else {
                connected[v].Add(u);
            }
        }
        
        int[] visited = new int[n + 1];
        long total = 0, optimal = 0;
        
        for (int i = 1; i <= n; i++) {
            if (count[i] > 0) {
                total += (long)count[i] * (count[i] - 1);
            }
            foreach (int j in connected[i]) {
                visited[j] = i;
            }
            foreach (int j in connected[i]) {
                foreach (int k in connected[j]) {
                    if (visited[k] == i) {
                        optimal++;
                    }
                }
            }
        }
        
        Console.WriteLine(total - 6 * optimal);
    }
}

File: 2018 ICPC Vietnam National Programming Contest/gottacatchemall.py
def main():
    import sys
    input = sys.stdin.read
    data = input().split()
    
    n = int(data[0])
    idx = 1
    
    junctions = []
    for _ in range(n - 1):
        u = int(data[idx]); v = int(data[idx + 1])
        idx += 2
        if u > v:
            u, v = v, u
        junctions.append((u, v))
    
    junctions = sorted(set(junctions))
    
    count = [0] * (n + 1)
    for u, v in junctions:
        count[u] += 1
        count[v] += 1
    
    connected = [[] for _ in range(n + 1)]
    for u, v in junctions:
        if count[u] != count[v]:
            if count[u] > count[v]:
                connected[u].append(v)
            else:
                connected[v].append(u)
        else:
            if u < v:
                connected[u].append(v)
            else:
                connected[v].append(u)
    
    visited = [0] * (n + 1)
    total = 0
    optimal = 0
    
    for i in range(1, n + 1):
        if count[i] > 0:
            total += count[i] * (count[i] - 1)
        for j in connected[i]:
            visited[j] = i
        for j in connected[i]:
            for k in connected[j]:
                if visited[k] == i:
                    optimal += 1
    
    print(total - 6 * optimal)

if __name__ == "__main__":
    main()